torch/csrc/autograd/engine.h

```diff
@@ -204,37 +253,50 @@ struct TORCH_API Engine {
 
   bool is_checkpoint_valid();
 
-  size_t ready_queue_size(at::Device device);
+  // initialize the thread local ready queue with the ready queue that is created
+  // elsewhere, i.e. thread_init, reentrant_thread_init, initialization in the
+  // distributed autograd engine, etc.
+  std::shared_ptr<ReadyQueue> init_local_ready_queue(std::shared_ptr<ReadyQueue> ready_queue = nullptr);
+
+  size_t ready_queue_size(const std::shared_ptr<GraphTask>& graph_task, at::Device device);
 
  protected:
   void compute_dependencies(Node* root, GraphTask& task);
   void evaluate_function(
       std::shared_ptr<GraphTask>& graph_task,
       Node* func,
       InputBuffer& inputs);
-  ReadyQueue& ready_queue(at::Device device);
-  ReadyQueue& ready_queue_by_index(int device_index);
-  void start_threads();
-  virtual void thread_init(int device);
+  ReadyQueue& ready_queue(
+      const std::shared_ptr<GraphTask>& graph_task,
+      at::Device device);
+  ReadyQueue& ready_queue_by_index(
+      const std::shared_ptr<GraphTask>& graph_task,
+      int device_index);
+  // start device threads (CUDA, XLA, etc.) in Engine,
+  // note that it does NOT start CPU thread.
+  void start_device_threads();
+  virtual void thread_init(int device, const std::shared_ptr<ReadyQueue>& ready_queue);
   virtual void thread_on_exception(
       std::shared_ptr<GraphTask> graph_task,
       const std::shared_ptr<Node>& fn,
       std::exception& e);
   virtual void thread_main(
       const std::shared_ptr<GraphTask>& task,
       bool reentrant_thread);
-  void reentrant_thread_init();
+  void reentrant_thread_init(const std::shared_ptr<ReadyQueue>& parent_ready_queue);
   void add_thread_pool_task(const std::weak_ptr<GraphTask>& graph_task);
   void set_device(int device);
-  void initialize_threads_pool();
+  void initialize_device_threads_pool();
 
   // Ensures ready_queues_ are initialized only once
-  std::once_flag start_threads_flag_;
+  std::once_flag start_device_threads_flag_;
   // Safe to read ready_queues_ without synchronization after intialization
-  std::vector<std::shared_ptr<ReadyQueue>> ready_queues_;
+  std::vector<std::shared_ptr<ReadyQueue>> device_ready_queues_;
   std::vector<std::function<void()>> final_callbacks_;
   // To protect reads and writes to final_callbacks_
   std::mutex post_callbacks_lock_;
+
+  std::mutex print_mutex_;
```

Comment:
What is this mutex used for?

Reply:
I think its a temporary mutex for logging that would be removed.